src/runtime/objmodel.cpp

```diff
@@ -1052,6 +1064,35 @@ Box* dataDescriptorInstanceSpecialCases(GetattrRewriteArgs* rewrite_args, const
         return runtimeCallInternal1(prop->prop_get, NULL, ArgPassSpec(1), obj);
     }
 
+    // Special case: data descriptor: getset descriptor
+    else if (descr->cls == getset_cls) {
+        BoxedGetsetDescriptor* getset_descr = static_cast<BoxedGetsetDescriptor*>(descr);
+
+        // TODO some more checks should go here
```

Comment:
Good question. I actually wrote this part ages ago and just dug it up last night... so I don't remember.

*Travis checks cpython*

```C
static int
descr_check(PyDescrObject *descr, PyObject *obj, PyObject **pres)
{
    if (obj == NULL) {
        Py_INCREF(descr);
        *pres = (PyObject *)descr;
        return 1;
    }
    if (!PyObject_TypeCheck(obj, descr->d_type)) {
        PyErr_Format(PyExc_TypeError,
                     "descriptor '%s' for '%s' objects "
                     "doesn't apply to '%s' object",
                     descr_name((PyDescrObject *)descr),
                     descr->d_type->tp_name,
                     obj->ob_type->tp_name);
        *pres = NULL;
        return 1;
    }
    return 0;
}
```

Looks like the getset descriptor saves what type it's supposed to apply to (`d_type`), and it checks that the type that the descriptor is on matches that.